// react-formalize/lib/components/Checkbox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _connectCheckbox = require('../hoc/connectCheckbox');

var _connectCheckbox2 = _interopRequireDefault(_connectCheckbox);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Checkbox = (0, _react.createClass)({
    displayName: 'Checkbox',
    propTypes: {
        name: _react.PropTypes.string.isRequired,
        checked: _react.PropTypes.bool.isRequired,
        onChange: _react.PropTypes.func.isRequired
    },
    getDefaultProps: function getDefaultProps() {
        return { type: 'text' };
    },
    render: function render() {
        return _react2.default.createElement('input', _extends({}, this.props, {
            type: 'checkbox' }));
    }
});

exports.default = (0, _connectCheckbox2.default)(Checkbox);
