// react-formalize/lib/components/Text.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _connectInput = require('../hoc/connectInput');

var _connectInput2 = _interopRequireDefault(_connectInput);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var types = ['date', 'datetime', 'datetime-local', 'email', 'month', 'number', 'password', 'tel', 'text', 'time', 'search', 'url', 'week'];

var Text = (0, _react.createClass)({
    displayName: 'Text',
    propTypes: {
        type: _react.PropTypes.oneOf(types),
        name: _react.PropTypes.string.isRequired,
        value: _react.PropTypes.string.isRequired,
        onChange: _react.PropTypes.func.isRequired
    },
    getDefaultProps: function getDefaultProps() {
        return { type: 'text' };
    },
    render: function render() {
        var onChange = this.props.onChange;

        return _react2.default.createElement('input', _extends({
            onCut: onChange,
            onPaste: onChange
        }, this.props));
    }
});

exports.default = (0, _connectInput2.default)(Text);
