// react-formalize/lib/components/TextArea.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _connectInput = require('../hoc/connectInput');

var _connectInput2 = _interopRequireDefault(_connectInput);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TextArea = (0, _react.createClass)({
    displayName: 'TextArea',
    propTypes: {
        name: _react.PropTypes.string.isRequired,
        value: _react.PropTypes.string.isRequired,
        onChange: _react.PropTypes.func.isRequired
    },
    render: function render() {
        var onChange = this.props.onChange;

        return _react2.default.createElement('input', _extends({
            onCut: onChange,
            onPaste: onChange
        }, this.props));
    }
});

exports.default = (0, _connectInput2.default)(TextArea);
