// sqlspaces/lib/space.js
/**
 *
 * User: giemza
 * Date: 09.02.12
 * Time: 20:30
 */

var Class = require('./class');

var Version = require('./version');

/**
 * Creates a new Space
 * @param {String} name The name of the space
 * @param {Number} id The id of the space
 * @param {Version} version the version of the space
 * @class
 */
module.exports = Class(/** @lends Space */{

    'constructor': function(name, id, version) {
        this.name = name;
        this.id = id;
        this.version = version;
    },

    'public': {

        /**
         * Returns the name
         * @returns {String} The name
         * @public
         */
        getName: function() {
            return this.name;
        },

        /**
         * Returns the id
         * @returns {Number} The id
         * @public
         */
        getID: function() {
            return this.id;
        },

        /**
         * Returns the version
         * @returns {Version} The version
         * @public
         */
        getVersion: function() {
            return this.version;
        },

        /**
         * Returns a JSON representation of the space
         * @returns {JSON} The JSON representation of the space
         * @public
         */
        toJSON: function() {
            return {
                'class': 'space',
                'id': this.fieldtype,
                'name': this.name,
                'version': this.version
            }
        },

        /**
         * Sets the properties of the space with an JSON object
         * @param {Object} data
         * @public
         */
        fromJSON: function(data) {
            this.id = data.id || null;
            this.name = data.name || null;
            var version = new Version();
            version.fromJSON(data.version || {});
        }
    }
});
